### PR TITLE
Add support for showing all package versions. Refs #175.

### PR DESCRIPTION
**Pull request by nickygerritsen**
In #175 there was a request to show the package versions. Since that is also a very useful feature for us, I have added this functionality.

Some notes / remarks:
* I added a package details page so one can view the versions here. We could later possibly add more here (like the readme). I made anonymous organization users have access to it.
* You can get to the details page both using the dropdown menu as well as clicking the package name.
* Does showing the clipboard button make sense? For `dev-*` versions this is what you want but for other versions you might want to add `^` or `~` in front of it. Maybe only let one copy the version itself as well as a separate button to copy the package name?
* I'm not sure if my logic in the Package and Version entity make sense?
* Perhaps we should add a `stable` boolean to version so we can show the `dev-*` versions in a separate table?
* I modified some tests to also test if we have the correct versions, but I'm not sure how/if we should test the reference as well? We could do it for the artifact case, since these ZIP's should have versions. For the local path one we can't assume anything since it uses local git info.

Some screenshots:
image
image

### CODE REVIEW

**Comment by nickygerritsen** [comment]
Not sure what I should do with the remaining phpstan errors, but those are related to one of my questions about how to handle this version logic

File: src/Query/User/PackageQuery/DbalPackageQuery.php [hunk 1]
```diff
@@ -103,6 +104,29 @@ public function getById(string $id): Option
         return Option::some($this->hydratePackage($data));
     }
 
+    /**
+     * @return Version[]
+     */
+    public function getVersions(string $packageId): array
```
Comment (akondas) [inline src/Query/User/PackageQuery/DbalPackageQuery.php:110]: it seems to me that it should be a paged list, because the quantity may grow, but at the moment it can stay that way (we can do it in a separate PR)

File: templates/organization/packages.html.twig [hunk 2]
```diff
@@ -16,7 +16,7 @@
         <thead>
             <tr>
                 <th>Name</th>
-                <th>Version</th>
+                <th>Latest version</th>
```
Comment (akondas) [inline templates/organization/packages.html.twig:19]: :+1:

File: src/Entity/Organization/Package.php [hunk 3]
```diff
@@ -246,4 +256,44 @@ private function setName(string $name): void
...
+     */
+    public function getVersion(string $versionString)
+    {
+        return $this->versions->filter(fn (Version $version) => $version->version() === $versionString)->first();
+    }
+
+    /**
+     * @param string[] $encounteredVersions
+     */
+    public function removeUnencounteredVersions(array $encounteredVersions): void
```
Comment (akondas) [inline src/Entity/Organization/Package.php:291]: what about put this logic to `syncSuccess` method? Now this method will be always called before `syncSuccess`, so i think we can add this `$encounteredVersions` param there.

File: templates/organization/package/details.html.twig [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+        <h3>Available versions</h3>
+
+        <table class="table">
+            <thead>
+            <tr>
+                <th>Version</th>
+                <th>Released</th>
+                <th>Reference</th>
+                <th></th>
```
Comment (akondas) [inline templates/organization/package/details.html.twig:38]: Maybe we can add `size` column here?

**Comment by akondas** [comment]
Looks great to me, thanks for your time :beers:.

File: src/Entity/Organization/Package.php (hunk 3 above)
Comment (nickygerritsen) [reply src/Entity/Organization/Package.php:291]: Maybe even pass the whole packages to syncsuccess and move the adding/removing logic fully there?

File: templates/organization/package/details.html.twig (hunk 4 above)
Comment (nickygerritsen) [reply templates/organization/package/details.html.twig:38]: Any idea how I can get the size easily? I could add a ->size to `distStorage` that returns the size of the archive?

File: src/Entity/Organization/Package.php (hunk 3 above)
Comment (akondas) [reply src/Entity/Organization/Package.php:291]: make sense

File: templates/organization/package/details.html.twig (hunk 4 above)
Comment (akondas) [reply templates/organization/package/details.html.twig:38]: Sounds good: `Storage -> size(Dist $dist): int`

File: src/Entity/Organization/Package.php (hunk 3 above)
Comment (nickygerritsen) [reply src/Entity/Organization/Package.php:291]: So given the below idea about storage: either we need to pass $diskStorage to `syncSuccess` or not move the whole packages there. I think the second option is better? Since otherwise we are passing services.

**Comment by nickygerritsen** [comment]
Also a colleague of my suggested that it might be more logical to add a copy button for the `composer require` command. Shall I do the following two things:
* Add a general copy button at the top that will `composer require <packagename>`; and
* Change the copy button for each such that it will `composer require <packagename>:<version>`

?

**Comment by akondas** [comment]
> Also a colleague of my suggested that it might be more logical to add a copy button for the `composer require` command. Shall I do the following two things:
> 
> * Add a general copy button at the top that will `composer require <packagename>`; and
> * Change the copy button for each such that it will `composer require <packagename>:<version>`
> 
> ?

This also sounds good :+1:

**Comment by nickygerritsen** [comment]
Fixes:
* Added pagination. Wasn't too hard so why not 🤷‍♂️ 
* Added size column
* Changed logic as requested, although adding/updating versions is still in the synchronize method, size we need the distStorage
* Updated the copy button things
* Added more tests to have at least the same coverage as before 🎉

File: src/Entity/Organization/Package.php (hunk 3 above)
Comment (akondas) [reply src/Entity/Organization/Package.php]: Passing these services is not bad. Generally, we will be doing a slight change of architecture towards a more layered (domain, application, infrastructure) to get away from this symfony style. At the moment it can stay as it is :+1:

**Comment by akondas** [comment]
@nickygerritsen After testing, I suggest you do one more thing: maybe we can add info when no versions are available:
Screenshot from 2020-07-07 08-11-13

```
To see available versions of the package, please perform synchronization again.
````
and button `Sync now`

Because currently all packages that have not been synchronized will have this page blank. The rest works fine. Good job :beers:

**Comment by akondas** [comment]
And what do you thing about something like this:
Screenshot from 2020-07-07 08-19-50
Light space compression. If it's better then I can add it to your PR.

**Comment by nickygerritsen** [comment]
I will do the first thing then you can do the second, looks much better.

**Comment by nickygerritsen** [comment]
image
Tada, like this?

**Comment by akondas** [comment]
:facepalm: Actually, we don't really need this extra table, after all, all available versions can be fetch from the metadata file `var/repo/organization/p/vendor/package.json`. You do not need to change anything during synchronization. In general, it facilitates the whole process. Just write a parser that extracts this data from this file. Sorry, I don't know why I didn't think about it before ... :thinking:

**Comment by nickygerritsen** [comment]
> 🤦 Actually, we don't really need this extra table, after all, all available versions can be fetch from the metadata file `var/repo/organization/p/vendor/package.json`. You do not need to change anything during synchronization. In general, it facilitates the whole process. Just write a parser that extracts this data from this file. Sorry, I don't know why I didn't think about it before ... 🤔

But will it be fast enough? Also pagination is then less useful since we need to read and `json_decode` the whole file anyway? Or do you still want it for client-side render speed?

I can try to write a parser for this data.

**Comment by akondas** [comment]
Actually, that's right, I saw 30MB files, that would kill rendering this page. Let's stay with what you managed to do now. If this does not work, we can always change this :wink:

**Comment by akondas** [comment]
This is my little UI refactor: https://github.com/Lets-Talk-NL/repman/pull/1/files :wink:

**Comment by nickygerritsen** [comment]
Nice, merged and squashed

**Comment by akondas** [comment]
Thanks @nickygerritsen
